orbotservice/src/main/java/org/torproject/android/service/circumvention/AutoConf.kt

```diff
@@ -0,0 +1,157 @@
+package org.torproject.android.service.circumvention
+
+import IPtProxy.IPtProxy
+import android.content.Context
+import java.net.Authenticator
+import java.net.PasswordAuthentication
+
+object AutoConf {
+
+    private const val MEEK_PARAMETERS = "url=https://1723079976.rsc.cdn77.org;front=www.phpmyadmin.net"
+
+    /**
+     * Tries to automatically configure Pluggable Transports, if the MOAT service decides, that in your country one is needed.
+     *
+     * @param country: An ISO 3166-1 Alpha 2 country code to force a specific country.
+     * 	 If not provided, the MOAT service will deduct a country from your IP address (preferred!)
+     * @param cannotConnectWithoutPt: Set to `true`, if you are sure, that a PT configuration *is needed*,
+     * 	 even though the MOAT service says, that in your country none is. In that case, a default configuration will be used.
+     * @return A transport to use and a list of custom bridges, if any could be evaluated.
+     *
+     * @throws MoatApi.MoatError or maybe other IO exceptions.
+     */
+    suspend fun `do`(context: Context, country: String? = null, cannotConnectWithoutPt: Boolean = false): Pair<Transport, List<String>>? {
+        var cannotConnectWithoutPt = cannotConnectWithoutPt
+
+        val done = fun(conf: Pair<Transport, List<String>>?): Pair<Transport, List<String>>? {
+            Transport.controller.stop(IPtProxy.MeekLite)
+            Authenticator.setDefault(null)
+
+            return conf
+        }
+
+        Transport.controller.start(IPtProxy.MeekLite, null)
+
+        val authenticator = object : Authenticator() {
+            override fun getPasswordAuthentication(): PasswordAuthentication? {
+                return PasswordAuthentication(MEEK_PARAMETERS, "\u0000".toCharArray())
+            }
+        }
+
+        Authenticator.setDefault(authenticator)
+
+        val api = MoatApi.getInstance(Transport.controller.port(IPtProxy.MeekLite).toInt())
+
+        // First, update built-ins.
+        if (BuiltInBridges.isOutdated(context)) {
+            val bridges = api.builtin()
+
+            if (!bridges.empty) {
+                bridges.store(context)
+
+                BuiltInBridges.invalidate()
+                BuiltInBridges.getInstance(context)
+            }
+        }
+
+        var response = api.settings(MoatApi.SettingsRequest(country))
+
+        val error = response.errors?.firstOrNull()
+        if (error != null) {
+            if (error.code == 404 /* Needs transport, but not the available ones */
+                || error.code == 406 /* no country from IP address */) {
+                cannotConnectWithoutPt = true
+            }
+            else {
+                throw error
```

Comment:
in the event that this error is caught somewhere down the line wouldn't meek still be running